File: src/foam/lib/query/StringParser.java
/**
 * @license
 * Copyright 2018 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

package foam.lib.query;

import foam.lib.json.ASCIIEscapeParser;
import foam.lib.json.UnicodeParser;
import foam.lib.parse.PStream;
import foam.lib.parse.Parser;
import foam.lib.parse.ParserContext;

public class StringParser implements Parser {
  public final static char ESCAPE = '\\';

  public StringParser() {
  }

  public PStream parse(PStream ps, ParserContext x) {

    if ( !ps.valid() )  return null;

    char delim = ' ';
    char lastc = delim;

    // TODO: use thread-local SB instead to avoid generating garbage
    StringBuilder sb = new StringBuilder();
    PStream lastPs=null;

    while ( ps.valid() ) {
      char c = ps.head();

      if ( c == '=' || c == '<' || c == '>' || c == '-' || c == ':' || c == ',' || c == ' ' || c == ')' ) {
        ps=lastPs;
        break;
      }
      
      if ( c == delim && lastc != ESCAPE ) break;
      
      lastPs=ps;
      PStream tail = ps.tail();

      if ( c == ESCAPE ) {
        char nextChar = ps.tail().head();
        Parser escapeSeqParser = null;

        if ( nextChar == 'u' ) {
          // TODO: make a constant
          escapeSeqParser = new UnicodeParser();
        } else if ( nextChar == 'n' ) {
          // TODO: make a constant
          escapeSeqParser = new ASCIIEscapeParser();
        }

        if ( escapeSeqParser != null ) {
          PStream escapePS = ps.apply(escapeSeqParser, x);
          if ( escapePS != null ) {
            sb.append(escapePS.value());
            tail = escapePS;
            c = ( (Character) escapePS.value() ).charValue();
          }
        }
      } else {
        sb.append(c);
      }

      ps = tail;
      lastc = c;
    }

    return ps.tail().setValue(sb.toString());
  }
}

File: src/foam/swift/parse/json/StringParser.js
/**
 * @license
 * Copyright 2017 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

foam.CLASS({
  package: 'foam.swift.parse.json',
  name: 'StringParser',
  implements: ['foam.swift.parse.parser.Parser'],
  axioms: [
    foam.pattern.Singleton.create()
  ],
  methods: [
    {
      name: 'parse',
      swiftCode: function() {/*
let escape_: Character = "\\"
let delim_ = ps!.head()
if delim_ != "\"" && delim_ != "'" { return nil }

var ps = ps!.tail()!;
var lastc = delim_;

var sb = ""
while ps.valid() {
  let c = ps.head()
  if c == delim_ && lastc != escape_ {
    break
  }
  if c != escape_ { sb.append(c) }
  lastc = c
  ps = ps.tail()!
}
return ps.tail()!.setValue(sb)
      */},
    },
  ]
});
